compiler/packages/babel-plugin-react-compiler/src/Transform/TransformFire.ts

```diff
@@ -0,0 +1,633 @@
+ *
+ * @param isUseEffect is necessary so we can keep track of when we should additionally insert
+ * useFire hooks calls.
+ */
+function visitFunctionExpressionAndPropagateFireDependencies(
+  fnExpr: FunctionExpression,
+  context: Context,
+  isUseEffect: boolean,
+): FireCalleesToFireFunctionBinding {
+  let withScope = isUseEffect
+    ? context.withUseEffectLambdaScope.bind(context)
+    : context.withFunctionScope.bind(context);
+
+  const calleesCapturedByFnExpression = withScope(() =>
+    replaceFireFunctions(fnExpr.loweredFunc.func, context),
+  );
+
+  /*
+   * Make a mapping from each dependency to the corresponding LoadLocal for it so that
+   * we can replace the loaded place with the generated fire function binding
+   */
+  const loadLocalsToDepLoads = new Map<IdentifierId, LoadLocal>();
+  for (const dep of fnExpr.loweredFunc.dependencies) {
+    const loadLocal = context.getLoadLocalInstr(dep.identifier.id);
+    if (loadLocal != null) {
+      loadLocalsToDepLoads.set(loadLocal.place.identifier.id, loadLocal);
+    }
+  }
+
+  const replacedCallees = new Map<IdentifierId, Place>();
+  for (const [
+    calleeIdentifierId,
+    loadedFireFunctionBindingPlace,
+  ] of calleesCapturedByFnExpression.entries()) {
+    /*
+     * Given the ids of captured fire callees, look at the deps for loads of those identifiers
+     * and replace them with the new fire function binding
+     */
+    const loadLocal = loadLocalsToDepLoads.get(calleeIdentifierId);
+    if (loadLocal == null) {
+      context.pushError({
+        loc: fnExpr.loc,
+        description: null,
+        severity: ErrorSeverity.Invariant,
+        reason:
+          '[InsertFire] No loadLocal found for fire call argument for lambda',
+        suggestions: null,
+      });
+      continue;
+    }
+
+    const oldPlaceId = loadLocal.place.identifier.id;
+    loadLocal.place = {
+      ...loadedFireFunctionBindingPlace.fireFunctionBinding,
+    };
+
+    replacedCallees.set(
+      oldPlaceId,
+      loadedFireFunctionBindingPlace.fireFunctionBinding,
+    );
+  }
+
+  // For each replaced callee, update the context of the function expression to track it
+  for (
+    let contextIdx = 0;
+    contextIdx < fnExpr.loweredFunc.func.context.length;
+    contextIdx++
+  ) {
+    const contextItem = fnExpr.loweredFunc.func.context[contextIdx];
+    const replacedCallee = replacedCallees.get(contextItem.identifier.id);
+    if (replacedCallee != null) {
+      fnExpr.loweredFunc.func.context[contextIdx] = replacedCallee;
+    }
+  }
+
+  context.mergeCalleesFromInnerScope(calleesCapturedByFnExpression);
+
+  return calleesCapturedByFnExpression;
+}
+
+function replaceFireFunctions(fn: HIRFunction, context: Context): void {
+  let hasRewrite = false;
+  for (const [, block] of fn.body.blocks) {
+    const rewriteInstrs = new Map<InstructionId, Array<Instruction>>();
+    const deleteInstrs = new Set<InstructionId>();
```

Comment:
the pattern we usually use for this is a nullable `nextInstructions` array, initialized by slicing up to the point where you need to start making edits. then after that you push instructions to keep/add, don't push instructions you want to delete

Reply:
that's an inconvenient approach for this pass specifically.

We insert the useFire calls before the load of the `useEffect`. If we were to wait until the point at which we know to generate them to insert, which is at the useEffect call, then we have to keep track of the index of the useEffect load index and insert the instructions there. This is easy to do with the InstructionIds because adding new instructions to the mapping does not shift the index of another useEffect load that occurs in the same block. If we push to the array, we'll need to additionally keep track of the number of added instructions so that we can account for the shifted offset in the nextInstructions array at which we should insert the useFire calls. If we naively insert the useFire calls right before the useEffect call then we introduce something like:
```
const effectLambda = () => {
  fireFunction();
}
const fireFunction = useFire(aFunction);
useEffect(effectLambda);
```

This code is legit, but the HIR representation for this is [not as simple](https://0xeac7-forget.vercel.app/#N4Igzg9grgTgxgUxALhAMygOzgFwJYSYAEAYhBABTBEBuAhgDZQJEC+AlEcADrFFyEwOImnJEAvEQqdxAPiIAjOjGkBuXkX6DhSmBKkz5AbXpMEAXXV8YCHLGKjK7K614hWQA) (requires a `DeclareContext HoistedConst`). It doesn't cause any of the tests to fail, but I'm not sure if that's really saying much.